File: backend/app/services/google_books.py
import httpx
from app.config import settings


async def google_books_search(q: str, start_index: int = 0) -> tuple[list[dict], int]:
    async with httpx.AsyncClient() as client:
        r = await client.get(
            "https://www.googleapis.com/books/v1/volumes",
            params={
                "q": q,
                "maxResults": "10",
                "startIndex": str(start_index),
                "key": settings.google_books_api_key,
            },
        )
        r.raise_for_status()
    data = r.json()
    items = [
        {
            "id": item.get("id"),
            "title": item.get("volumeInfo", {}).get("title"),
            "authors": item.get("volumeInfo", {}).get("authors", []),
            "thumbnail": item.get("volumeInfo", {}).get("imageLinks", {}).get("thumbnail"),
            "publishedDate": item.get("volumeInfo", {}).get("publishedDate"),
            "description": item.get("volumeInfo", {}).get("description"),
            "pageCount": item.get("volumeInfo", {}).get("pageCount"),
            "categories": item.get("volumeInfo", {}).get("categories", []),
            "infoLink": item.get("volumeInfo", {}).get("infoLink"),
        }
        for item in data.get("items", [])
    ]
    return items, data.get("totalItems", 0)


async def run_tool(name: str, args: dict, start_index: int = 0) -> tuple[list[dict], int]:
    match name:
        case "get_books_by_title":
            return await google_books_search(f"intitle:{args['title']}", start_index)
        case "get_books_by_author":
            return await google_books_search(f"inauthor:{args['author']}", start_index)
        case "get_books_by_isbn":
            return await google_books_search(f"isbn:{args['isbn']}", start_index)
        case _:
            raise ValueError(f"Unknown tool: {name}")

File: api/src/services/googleBooks.ts
import { config } from '../config/index.js';

export interface BookItem {
  id: string;
  title: string;
  authors: string[];
  thumbnail?: string;
  publishedDate?: string;
  description?: string;
  pageCount?: number;
  categories: string[];
  infoLink?: string;
}

export interface GoogleBooksResult {
  items: BookItem[];
  totalItems: number;
}

async function search(q: string, startIndex: number): Promise<GoogleBooksResult> {
  const url = new URL('https://www.googleapis.com/books/v1/volumes');
  url.searchParams.set('q', q);
  url.searchParams.set('maxResults', '10');
  url.searchParams.set('startIndex', String(startIndex));
  url.searchParams.set('key', config.GOOGLE_BOOKS_API_KEY);

  const res = await fetch(url.toString());
  if (!res.ok) throw new Error(`Google Books API error: ${res.status}`);

  const data = (await res.json()) as { totalItems?: number; items?: any[] };

  const items: BookItem[] = (data.items ?? []).map((item: any) => {
    const vi = item.volumeInfo ?? {};
    return {
      id: item.id,
      title: vi.title ?? '',
      authors: vi.authors ?? [],
      categories: vi.categories ?? [],
      ...(vi.imageLinks?.thumbnail && { thumbnail: vi.imageLinks.thumbnail }),
      ...(vi.publishedDate && { publishedDate: vi.publishedDate }),
      ...(vi.description && { description: vi.description }),
      ...(vi.pageCount && { pageCount: vi.pageCount }),
      ...(vi.infoLink && { infoLink: vi.infoLink }),
    };
  });

  return { items, totalItems: data.totalItems ?? 0 };
}

export async function searchByTitle(title: string, startIndex: number): Promise<GoogleBooksResult> {
  return search(`intitle:${title}`, startIndex);
}

export async function searchByAuthor(author: string, startIndex: number): Promise<GoogleBooksResult> {
  return search(`inauthor:${author}`, startIndex);
}

export async function searchByIsbn(isbn: string, startIndex: number): Promise<GoogleBooksResult> {
  return search(`isbn:${isbn}`, startIndex);
}
